src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp

```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+// [out] pdwAge - PDB's age to be placed here
+// [out] wszPath - PDB's build path to be placed here
+// [in] cchPath - Number of wide characters allocated in wszPath, including NULL terminator
+//
+// This is a simplification of similar code in CLR's GetCodeViewInfo
+// in eventtrace.cpp.
+void PalGetPDBInfo(HANDLE hOsHandle, GUID * pGuidSignature, _Out_ uint32_t * pdwAge, _Out_writes_z_(cchPath) WCHAR * wszPath, int32_t cchPath)
+{
+    // Zero-init [out]-params
+    ZeroMemory(pGuidSignature, sizeof(*pGuidSignature));
+    *pdwAge = 0;
+    if (cchPath <= 0)
+        return;
+    wszPath[0] = L'\0';
+
+    BYTE *pbModule = (BYTE*)hOsHandle;
+
+    IMAGE_NT_HEADERS const * pNtHeaders = (IMAGE_NT_HEADERS*)(pbModule + ((IMAGE_DOS_HEADER*)hOsHandle)->e_lfanew);
+    IMAGE_DATA_DIRECTORY const * rgDataDirectory = NULL;
+    if (pNtHeaders->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC)
+        rgDataDirectory = ((IMAGE_OPTIONAL_HEADER32 const *)&pNtHeaders->OptionalHeader)->DataDirectory;
+    else
+        rgDataDirectory = ((IMAGE_OPTIONAL_HEADER64 const *)&pNtHeaders->OptionalHeader)->DataDirectory;
+
+    IMAGE_DATA_DIRECTORY const * pDebugDataDirectory = &rgDataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG];
+
+    // In Redhawk, modules are loaded as MAPPED, so we don't have to worry about dealing
+    // with FLAT files (with padding missing), so header addresses can be used as is
+    IMAGE_DEBUG_DIRECTORY const *rgDebugEntries = (IMAGE_DEBUG_DIRECTORY const *) (pbModule + pDebugDataDirectory->VirtualAddress);
+    DWORD cbDebugEntries = pDebugDataDirectory->Size;
+    if (cbDebugEntries < sizeof(IMAGE_DEBUG_DIRECTORY))
+        return;
+
+    // Since rgDebugEntries is an array of IMAGE_DEBUG_DIRECTORYs, cbDebugEntries
+    // should be a multiple of sizeof(IMAGE_DEBUG_DIRECTORY).
+    if (cbDebugEntries % sizeof(IMAGE_DEBUG_DIRECTORY) != 0)
+        return;
+
+    // CodeView RSDS debug information -> PDB 7.00
+    struct CV_INFO_PDB70 
+    {
+        DWORD          magic; 
+        GUID           signature;       // unique identifier 
+        DWORD          age;             // an always-incrementing value 
+        _Field_z_ char  path[MAX_PATH];  // zero terminated string with the name of the PDB file 
+    };
+
+    // Temporary storage for a CV_INFO_PDB70 and its size (which could be less than
+    // sizeof(CV_INFO_PDB70); see below).
+    struct PdbInfo
+    {
+        CV_INFO_PDB70 *     m_pPdb70;
+        ULONG               m_cbPdb70;
+    };
+
+    // Grab module bounds so we can do some rough sanity checking before we follow any
+    // RVAs
+    uint8_t * pbModuleLowerBound = NULL;
+    uint8_t * pbModuleUpperBound = NULL;
+    PalGetModuleBounds(hOsHandle, &pbModuleLowerBound, &pbModuleUpperBound);
+
+    // Iterate through all debug directory entries. The convention is that debuggers &
+    // profilers typically just use the very last IMAGE_DEBUG_TYPE_CODEVIEW entry.  Treat raw
+    // bytes we read as untrusted.
```

Comment:
```suggestion
    // profilers typically just use the very last IMAGE_DEBUG_TYPE_CODEVIEW entry.
```
It does not make sense to not trust our own binary.